sentry_sdk/sessions.py

```diff
@@ -0,0 +1,247 @@
+        # type: (...) -> None
+        self._hub = weakref(hub)
+        if sid is None:
+            sid = uuid.uuid4()
+        if started is None:
+            started = datetime.utcnow()
+        if status is None:
+            status = "ok"
+        self.status = status
+        self.did = None  # type: Optional[str]
+        self.started = started
+        self.release = None  # type: Optional[str]
+        self.environment = None  # type: Optional[str]
+        self.duration = None  # type: Optional[float]
+        self.user_agent = None  # type: Optional[str]
+        self.ip_address = None  # type: Optional[str]
+        self.errors = 0
+
+        self.update(
+            sid=sid,
+            did=did,
+            timestamp=timestamp,
+            duration=duration,
+            release=release,
+            environment=environment,
+            user_agent=user_agent,
+            ip_address=ip_address,
+            errors=errors,
+        )
+
+    def update(
+        self,
+        sid=None,  # type: Optional[Union[str, uuid.UUID]]
+        did=None,  # type: Optional[str]
+        timestamp=None,  # type: Optional[datetime]
+        duration=None,  # type: Optional[float]
+        status=None,  # type: Optional[SessionStatus]
+        release=None,  # type: Optional[str]
+        environment=None,  # type: Optional[str]
+        user_agent=None,  # type: Optional[str]
+        ip_address=None,  # type: Optional[str]
+        errors=None,  # type: Optional[int]
+    ):
+        # type: (...) -> None
+        hub = self._hub()
+        options = hub.client.options if hub and hub.client else None
+
+        if sid is not None:
+            self.sid = _make_uuid(sid)
+        if did is not None:
+            self.did = str(did)
+        if timestamp is None:
+            timestamp = datetime.utcnow()
+        self.timestamp = timestamp
+        if duration is not None:
+            self.duration = duration
+        if release is None and options:
+            release = options["release"]
+        if release is not None:
+            self.release = release
+        if environment is None and options:
+            environment = options["environment"]
+        if environment is not None:
+            self.environment = environment
+        if ip_address is not None:
+            self.ip_address = ip_address
+        if user_agent is not None:
+            self.user_agent = user_agent
+        if errors is not None:
+            self.errors = errors
+
+        if status is not None:
+            self.status = status
+
+    def close(
+        self, status=None  # type: Optional[SessionStatus]
+    ):
+        # type: (...) -> Any
+        if status is None and self.status == "ok":
+            status = "exited"
+        if status is not None:
+            self.update(status=status)
+
+        # sessions get flushed in this SDK only when they close (single
+        # session update).
+        hub = self._hub()
+        if hub is not None:
```

Comment:
What about `Scope.closeSession` atomically changes the session state to ended, set the session field to null and returns the instance? The caller (Hub) would be responsible to flush it. 

I'm just trying to avoid the double dispatch with weakref here but maybe I need to read this better.

Also because if we do `withScope(s -> s.session.close())` now we close (and flush) but we end up with a finished/flushed session bound to the scope.

Reply:
Experimented with `hub.startSession` where no Hub is given to the Session.
Lots of stuff missing so likely just missing your point here:

https://github.com/getsentry/sentry-android/pull/281/files#diff-799727108cf81a8a2e3bd30a102ef33aR163